gcc/config/riscv/riscv.cc

```diff
@@ -3728,6 +3728,74 @@ riscv_pass_fpr_pair (machine_mode mode, unsigned regno1,
 				   GEN_INT (offset2))));
 }
 
+/* Use the TYPE_SIZE to distinguish the type with vector_size attribute and
+   intrinsic vector type.  Because we can't get the decl for the params.  */
+
+static bool
+riscv_arg_has_vector_size_attribute (const_tree type)
+{
+  tree size = TYPE_SIZE (type);
+  if (size && TREE_CODE (size) == INTEGER_CST)
+    return true;
+
+  /* For the data type like vint32m1_t, the size code is POLY_INT_CST.  */
+  return false;
+}
+
+static bool
+riscv_arg_has_vector (const_tree type)
+{
+  bool is_vector = false;
+
+  switch (TREE_CODE (type))
+    {
+    case RECORD_TYPE:
+      if (!COMPLETE_TYPE_P (type))
+	break;
+
+      for (tree f = TYPE_FIELDS (type); f; f = DECL_CHAIN (f))
+	if (TREE_CODE (f) == FIELD_DECL)
+	  {
+	    if (!TYPE_P (TREE_TYPE (f)))
+	      break;
+
+	    /* If there's vector_size attribute, ignore it.  */
+	    if (VECTOR_TYPE_P (TREE_TYPE (f)))
+	      is_vector = !riscv_arg_has_vector_size_attribute (type);
+	    else
+	      is_vector = riscv_arg_has_vector (TREE_TYPE (f));
+	  }
+
+      break;
+
+    case VECTOR_TYPE:
+      is_vector = !riscv_arg_has_vector_size_attribute (type);
+      break;
+
+    default:
+      is_vector = false;
+      break;
+    }
+
+  return is_vector;
+}
+
+/* Pass the type to check whether it's a vector type or contains vector type.
+   Only check the value type and no checking for vector pointer type.  */
+
+static void
+riscv_pass_in_vector_p (const_tree type)
+{
+  static int warned = 0;
+
+  if (type && riscv_arg_has_vector (type) && !warned)
+    {
+      warning (OPT_Wpsabi, "ABI for the vector type is currently in experimental"
+	       "stage and may changes in the upcoming version of GCC.");
```

Comment:
will the message become something like `in experimentalstage and ...` instead of `in experimental stage and ...` after joining?

Reply:
Yes. You're right. Will modify it in the fix commit.

Reply:
Cool, thank you!